pkg/kube/pod.go

```diff
@@ -114,16 +119,71 @@ func GetPodLogs(ctx context.Context, cli kubernetes.Interface, namespace, name s
 
 // WaitForPodReady waits for a pod to exit the pending state
 func WaitForPodReady(ctx context.Context, cli kubernetes.Interface, namespace, name string) error {
-	err := poll.Wait(ctx, func(ctx context.Context) (bool, error) {
+	timeoutCtx, waitCancel := context.WithTimeout(ctx, podReadyWaitTimeout)
+	defer waitCancel()
+	err := poll.Wait(timeoutCtx, func(ctx context.Context) (bool, error) {
 		p, err := cli.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
 		if err != nil {
 			return false, err
 		}
+
+		// check if nodes are up and available
+		n := strings.Split(p.Spec.NodeName, "/")
+		if n[0] != "" {
+			node, err := cli.CoreV1().Nodes().Get(n[0], metav1.GetOptions{})
+			if err != nil {
+				return false, errors.Wrapf(err, "Failed to get node %s", n[0])
+			}
+			if !IsNodeReady(node) || !IsNodeSchedulable(node) {
+				return false, errors.Errorf("Node %s is currently not ready/schedulable", n[0])
+			}
+		}
+
+		// check for memory or resource issues
+		if p.Status.Phase == v1.PodPending {
+			if p.Status.Reason == "OutOfmemory" || p.Status.Reason == "OutOfcpu" {
+				return false, errors.Errorf("Pod stuck in pending state, reason: %s", p.Status.Reason)
+			}
+		}
+
+		// check if pvc and pv are up and ready to mount
+		if err := checkPVCAndPVStatus(p, cli, namespace); err != nil {
+			return false, err
+		}
+
 		return p.Status.Phase != v1.PodPending && p.Status.Phase != "", nil
 	})
 	return errors.Wrapf(err, "Pod did not transition into running state. Namespace:%s, Name:%s", namespace, name)
 }
 
+func checkPVCAndPVStatus(p *v1.Pod, cli kubernetes.Interface, namespace string) error {
+	for _, vol := range p.Spec.Volumes {
+		if vol.VolumeSource.PersistentVolumeClaim != nil {
+			pvcName := vol.VolumeSource.PersistentVolumeClaim.ClaimName
+			pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(pvcName, metav1.GetOptions{})
+			if err != nil {
+				return errors.Wrapf(err, "Failed to get pvc %s", pvcName)
```

Comment:
[nit] Comparing with other logs. Some have caps and some don't
```suggestion
				return errors.Wrapf(err, "Failed to get PVC %s", pvcName)
```